lib/DxilPIXPasses/DxilShaderAccessTracking.cpp

```diff
@@ -385,49 +385,53 @@ bool DxilShaderAccessTracking::runOnModule(Module &M)
           FOS << "ShouldAssumeDsvAccess";
         }
       }
-      IRBuilder<> Builder(DM.GetEntryFunction()->getEntryBlock().getFirstInsertionPt());
-
-      unsigned int UAVResourceHandle = static_cast<unsigned int>(DM.GetUAVs().size());
-
-      // Set up a UAV with structure of a single int
-      SmallVector<llvm::Type*, 1> Elements{ Type::getInt32Ty(Ctx) };
-      llvm::StructType *UAVStructTy = llvm::StructType::create(Elements, "class.RWStructuredBuffer");
-      std::unique_ptr<DxilResource> pUAV = llvm::make_unique<DxilResource>();
-      pUAV->SetGlobalName("PIX_CountUAVName");
-      pUAV->SetGlobalSymbol(UndefValue::get(UAVStructTy->getPointerTo()));
-      pUAV->SetID(UAVResourceHandle);
-      pUAV->SetSpaceID((unsigned int)-2); // This is the reserved-for-tools register space
-      pUAV->SetSampleCount(1);
-      pUAV->SetGloballyCoherent(false);
-      pUAV->SetHasCounter(false);
-      pUAV->SetCompType(CompType::getI32());
-      pUAV->SetLowerBound(0);
-      pUAV->SetRangeSize(1);
-      pUAV->SetKind(DXIL::ResourceKind::RawBuffer);
-
-      auto pAnnotation = DM.GetTypeSystem().GetStructAnnotation(UAVStructTy);
-      if (pAnnotation == nullptr) {
-
-          pAnnotation = DM.GetTypeSystem().AddStructAnnotation(UAVStructTy);
-          pAnnotation->GetFieldAnnotation(0).SetCBufferOffset(0);
-          pAnnotation->GetFieldAnnotation(0).SetCompType(hlsl::DXIL::ComponentType::I32);
-          pAnnotation->GetFieldAnnotation(0).SetFieldName("count");
-      }
 
-      ID = DM.AddUAV(std::move(pUAV));
+      for (llvm::Function & F : M.functions()) {
```

Comment:
Does this pass get run before linking library targets? In that case, you'd get more than just entry points.

Reply:
Thanks for pointing that out. PIX (as yet) doesn't run this pass on libraries. Just on non-raytracing PSOs, which (currently) implies single-shader post-link binaries. I'm pretty sure this whole pass will need to be updates if/when PIX gets around to applying it in the raytrace case anyway. That output UAV will have to be much more general and/or we'll need more UAVs.